api/commands/format.ts

```diff
@@ -0,0 +1,182 @@
+          .map(child => {
+            if (typeof child !== "string") {
+              const placeholderId = crypto.randomBytes(16).toString("hex");
+              placeholders[placeholderId] = child;
+              return placeholderId;
+            }
+            return child;
+          })
+          .join("");
+        return Object.keys(placeholders).reduce(
+          (partiallyFormattedSql, placeholderId) =>
+            formatPlaceholderInSqlx(
+              placeholderId,
+              placeholders[placeholderId],
+              partiallyFormattedSql
+            ),
+          sqlFormatter.format(unformattedPlaceholderSql) as string
+        );
+      })
+      .join("\n\n---\n\n")
+  );
+}
+
+function getIndividualSqlxStatements(nodeContents: Array<string | ISyntaxTreeNode>) {
+  const sqlxStatements: Array<Array<string | ISyntaxTreeNode>> = [[]];
+  nodeContents.forEach(child => {
+    if (typeof child !== "string" && child.startTokenType.endsWith("_statementSeparator")) {
+      sqlxStatements.push([]);
+    } else {
+      sqlxStatements[sqlxStatements.length - 1].push(child);
+    }
+  });
+  return sqlxStatements;
+}
+
+function formatJavaScript(text: string) {
+  return jsBeautify.js(text, JS_BEAUTIFY_OPTIONS);
+}
+
+function formatPlaceholderInSqlx(
+  placeholderId: string,
+  placeholderSyntaxNode: ISyntaxTreeNode,
+  sqlx: string
+) {
+  const wholeLineContainingPlaceholderId = getWholeLineContainingPlaceholderId(placeholderId, sqlx);
+  // If the placeholder is the first non-whitespace on this line, use its current indentation.
+  const indentSize = wholeLineContainingPlaceholderId.trimLeft().startsWith(placeholderId)
+    ? wholeLineContainingPlaceholderId.length - wholeLineContainingPlaceholderId.trimLeft().length
+    : 0;
+  const formattedChild = formatChildSyntaxTreeNode(placeholderSyntaxNode, indentSize);
+
+  // Figure out what text we need to replace.
+  // If the placeholder is the first non-whitespace text on this line, replace the placeholder
+  // and all preceding whitespace (on that line). Otherwise, replace just the placeholder.
+  const leftOfPlaceholder = wholeLineContainingPlaceholderId.slice(
+    0,
+    wholeLineContainingPlaceholderId.indexOf(placeholderId)
+  );
+  const textToReplace =
+    leftOfPlaceholder.trim().length === 0
+      ? wholeLineContainingPlaceholderId.slice(0, leftOfPlaceholder.length + placeholderId.length)
+      : placeholderId;
+  return sqlx.replace(textToReplace, formattedChild);
+}
+
+function formatChildSyntaxTreeNode(node: ISyntaxTreeNode, jsIndent: number): string {
+  if (node.startTokenType.endsWith("_startJsPlaceholder")) {
+    return formatJavaScriptPlaceholder(node, jsIndent);
+  }
+  if (node.startTokenType.endsWith("_startJs") || node.startTokenType.endsWith("_startConfig")) {
+    return `\n\n${formatJavaScript(concatenateSyntaxTreeContents(node))}\n\n`;
+  }
+  // This node must be an "inner" SQL block, e.g. "pre_operations { ... }",
```

Comment:
Should we be explicit here? We will inevitably add new types that might not wanted to be treated like this. Might be better to just leave such blocks as they were

Reply:
done!